include/cetl/span.h

```diff
@@ -1 +1,772 @@
+    // | Member Types
+    // +----------------------------------------------------------------------+
+    using element_type     = T;                                 //!< The element type over which the span operates
+    using value_type       = typename std::remove_cv<T>::type;  //!< The non-cv type of elements
+    using pointer          = T*;                                //!< The element pointer type
+    using const_pointer    = const T*;                          //!< A constant pointer to the element type
+    using reference        = T&;                                //!< A reference type to the element
+    using const_reference  = const T&;                          //!< A constant reference type to the element
+    using size_type        = std::size_t;                       //!< Used to track the size in bytes
+    using difference_type  = std::ptrdiff_t;                    //!< Type used to compare two locations in the span.
+    using iterator         = pointer;                           //!< Type for iterating through the span.
+    using reverse_iterator = std::reverse_iterator<iterator>;   //!< Reverse iterator type.
+
+    // +----------------------------------------------------------------------+
+    // | Required Properties
+    // +----------------------------------------------------------------------+
+    static_assert(not std::is_pointer<T>::value, "Can not be a pointer type");
+    static_assert(sizeof(element_type) != 0u, "Must have non zero element size");
+
+    // +----------------------------------------------------------------------+
+    ///@{ @name Constructors
+    // +----------------------------------------------------------------------+
+
+    /// Default constructor
+    /// @tparam DeducedExtent   The Extent deduced by the compiler.
+    /// @tparam type            SFNAE enablement of this constructor only where the extent > 0.
+    /// @see std::span::span()
+    template <std::size_t DeducedExtent = Extent, typename std::enable_if<(DeducedExtent == 0), bool>::type = true>
+    constexpr span() noexcept
+        : data_{nullptr}
+    {
+    }
+
+    /// Creates a span starting at an element for a given length.
+    /// The span's view into the data starts at the element pointed to by the first pointer. The size of the
+    /// span is provided by the count where it is undefined behavior to provide a count value that is not the
+    /// same as span::extent.
+    /// @param first The first element in the span.
+    /// @param count The number of elements in the span.
+    /// @see std::span::span()
+    explicit constexpr span(iterator first, size_type count)
+        : data_{first}
+    {
+        (void) count;  // Count isn't actually used in static spans.
+        CETL_DEBUG_ASSERT(count == extent,
+                          "Undefined behavior: Constructing a fixed span where the Extent parameter is different from "
+                          "the count passed into this constructor.");
```

Comment:
This is about all you can do for static spans but people will complain. Also this will be problematic to detect at runtime on embedded targets (as me how I know).

Reply:
details?

Reply:
misconfiguring the compile time size and the runtime size, causing the assert can be difficult to track down and ultimately a weird problem to choose to allow. My personal preference is to make `span` an interface which is either implemented by a FixedSpan or a DynamicSpan.

Reply:
Based on the specification this constructor:
![image](https://user-images.githubusercontent.com/19180861/224168460-7d80c243-53cd-44f1-8467-1c6a28f2ff89.png)
given the required preconditions:
![image](https://user-images.githubusercontent.com/19180861/224168604-57e982c9-aa1b-4b7d-918b-7f0f6ed6a085.png)
and that `count` has to be evaluated at runtime, I'm not sure there is any other recourse here.